### PR TITLE
refactor: customise sanityAlgolia webhook helpers

### LINKED ISSUE #997: [S76.1]: Hits update for bundles
### Link to Epic

Epic: https://github.com/Open-Study-College/osc/issues/993

### Description

n/a

### Linked Tasks

- [ ] #998 
- [ ] #1444

### LINKED ISSUE #1005: Create Web Hooks
### Link task to story

Story: https://github.com/Open-Study-College/osc/issues/1004

### Description

n/a

### LINKED ISSUE #1272: Adding Product Variants and any Missing Metafields into Sanity/Algolia Script
### Link task to story

Story: #1004

### Description

We need to adapt the sanity to algolia webhook to include Product Variants and any missing Metafields such as:
- Awards
- Awarding Bodies
- Course Duration
- Product Categories (currently define as Product Tags e.g. `Child_col:Name of Child Category`, `Parent_col:Name of Category`)

The sanity webhook script will effectively pull missing metafield data from shopify and stitch it together to push into Algolia

### LINKED ISSUE #1273: Sorting Faceting Data
### Link task to story

Story: #1004

### Description

Off the back of the work to add product variants and missing metafields to import script, we now need to sort the faceting data which includes adding the searchable attributes and looking into how the price/price range will now work.

### LINKED ISSUE #1274: Adding Collections Facet
### Link task to story

Story: #1004

### Description

Now, that collections (or product categories) are assigned to metafields, we need to now add as a filterable facet.

### LINKED ISSUE #1443: Add 'lowest monthly repayment' into Algolia
### Link task to story

Story: #1020 

### Description

Add lowest monthly repayment metafield into the sanityAlgolia webhook to sync this data across to Algolia.

### LINKED ISSUE #1444: Add product_type to Algolia
### Link task to story

Story: #997

### Description

The `product_type` from Shopify needs adding to Algolia so we can update whether courses are single or package

### LINKED ISSUE #1448: Setup indexes in Algolia Production
### Link task to story

Story: #997 

### Description

We need to replicate all the dev indexes in our production Algolia, set up all the searchable attributes and facets. 

1) Set up webhook in Production Shopify to sync the collections to Production Algolia
2) Run the SanityAlgolia fullSync against Production Algolia
3) Create all the Replicas - Grouped by id and then the sort by lists, e.g. price asc/desc
4) Set up all the searchable attributes - must be done against ALL the replicas
5) Set up the Facets - must be done against ALL the replicas

### LINKED ISSUE #1450: Add Collections into Algolia Indexes
### Link task to story

Story: 

### Description

Need to work out a way of synchronising collections to Algolia correctly. 

Once this is done, they should be added to the SanityAlgolia webhook. Need to make sure that any change to a collection in Sanity will then automatically update all of the correct products that are linked to the changed collection

### LINKED ISSUE #1521: Algolia Webhook failing due to rate limit issues with Shopify API
### Description

When I try to update all products it fails because the Shopify API timesout

### Link to Reproduction

localhost:2000

### Steps to reproduce

Run sanityAlgolia webhook as a fullsync

### Release Version

v0.177.0

### Browser

_No response_

### Operating System

- [ ] macOS
- [ ] Windows
- [ ] Linux

### Additional Information

_No response_

### LINKED ISSUE #1582: Add namedTags and compareAtPrices to Algolia
### Link task to story

Story: #1004

### Description

We need to add named tags and compare at prices to Algolia.

These existed in the original import of products from Shopify but wasn't something that had been integrated when moving over to using SanityAlgolia helpers. They are required for sale price information to display correctly on the /search page.

### PR DESCRIPTION
**Pull request by mattwendzina**
### 🧹 Please check off the following before submitting:

- [X] Close issues
- [X] Request reviewers
- [X] Assign yourself
- [X] Add labels (e.g. `build:osc-ecommerce`)
- [X] Add project
- [x] Set project status
- [X] Add milestone

---

## ☑ List all github issues that this PR will close:

- Closes #997 
- Closes #1005 
- Closes #1272
- Closes #1273 
- Closes #1274 
- Closes #1443 
- Closes #1444 
- Closes #1448 
- Closes #1450 
- Closes #1582
- Closes #1521

## 📝 Description

This deals specifically with the updates to the SanityAlgolia helper functions for the Sanity Webhook. I've created a new PR because I wanted to separate out a few different pieces of the work as it was becoming a bit unwieldy.

I have also refactored quite a bit of code into helper functions because there was increasingly a lot of code repetition. I also wanted to have helper functions for whenever we reach out to the different APIs (Sanity, Algolia, Shopify) so we can handle errors better if there any network/api issues that could cause data to be incorrect or incomplete.

Because this PR now only deals with the Sanity Algolia helper functions and syncing the data, you won't see the updates pull through on the front end from this PR.

There will be proceeding PRs that deal with cascading all the data changes from Algolia into the front end.

## ⛳️ Current behaviour (updates)

The Sanity Webhook will only update pages and blogs

## 🚀 New behaviour

These updates customise the helpers provided by the SanityAlgolia package so that we can update all the product variants. Also included following updates:
- Merges and syncs all of the necessary meta data that is required on the front end. 
- Deals with logic around creating information on package courses - the maximum saving, as well as adding information to all products that are related to a package and the savings. It also handles some complex logic around how we effectively ensure Algolia is kept in sync when package updates are made, or product updates are made on products that are part of a package.

## 💣 Is this a breaking change (Yes/No):

No

## 🧾 Final Checklist

- [ ] I've considered possible cross-platform impacts (Mac, Linux, Windows)
- [ ] I've considered possible cross-device impacts (iOS, Android)
- [ ] I've considered possible cross-browser impacts (Chrome, Safari, Edge, Firefox)
- [ ] I've added tests to cover my changes (Unit, E2E, Integration, Contract, Functional, Acceptance etc)
- [ ] I've added or updated any documentation.

## 📝 Additional Information

I've left a few console logs in, this is just so that if any issues crop up once this goes into production and marketing are making updates we can hopefully identify the issues a bit more quickly.

I'm copying and pasting the information from the `Update Sanity Algolia Webhook` PR here on how to run updates:

This could do with a few people testing to see if it can be broken or there are any problems. We should identify any issues now to reduce the risk of stale or inaccurate search data once we go live.

I have created yet another (😅) application in Algolia to most accurately represent what I think we would want when we go to production. You should have been given permissions to be able to access it.

You will need all of the following env vars. If you don't have any of these ask me and I can pass them on.

#Algolia
ALGOLIA_APP_ID=""
ALGOLIA_API_KEY_SEARCH=""
ALGOLIA_API_KEY_ADMIN=""
ALGOLIA_PRIMARY_PRODUCTS_INDEX=""
ALGOLIA_PRIMARY_COLLECTIONS_INDEX=""
ALGOLIA_PRODUCTS_INDEX_GROUPED_BY_ID=""
ALGOLIA_DEV_ALL_RECORDS_INDEX = ""

...
#Sanity
SANITY_STUDIO_API_DATASET=""
SANITY_WEBHOOK_SECRET=""
...
# Hydrogen
All hydrogen dev credentials - if you don't have these ask me or Steve
In order to test the webhook you will need a tool such as localtunnel (very quick and easy to set up/use - https://github.com/localtunnel/localtunnel). This creates a public url which is required to use the Sanity webhooks.
Once your tunnel is running, go to 'Manage Project' in Sanity and select the API tab. You should see an existing webhook called SanityAlgolia webhook - click 'Edit Webhook' (make sure to choose the DEVELOPMENT one!) You will need to update the URL to the one provided by your tunnel. Makes sure to keep the /actions/sanityAlgolia parts of the URL, e.g. yourTunnelUrl/actions/sanityAlgolia. It's probably worth dropping a message in Discord just to check no one else is testing/set up the URL to point to their tunnel.

Once this is set up, you should be able to test the webhook. At this point any changes to products in Shopify (Dev store) and any updates to posts/pages/products in Sanity (staging) should activate the webhook. Once you make a change, such as updating/deleteing/creating a variant or product in Shopify you should then check to see if the changes have been synchronised to Algolia. You can also check the logs in the webhook to see if they have been successful.

### Summary by CodeRabbit

- New Feature: Added integration with Algolia for enhanced search functionality. This includes syncing data from Sanity and Shopify to Algolia.
- New Feature: Introduced new GraphQL queries for fetching product, variant, blog, and post information from Sanity and Shopify.
- Refactor: Updated the `AlgoliaHit` interface and added several new types and interfaces related to Algolia indexing and webhooks.
- Chore: Added a new environment variable `ALGOLIA_PRODUCTION_ALL_RECORDS_INDEX` across various configurations for specifying the Algolia index in the production environment.

### CODE REVIEW

**Review by stevenbister — CHANGES_REQUESTED** [review]
Blimey mate well done 😅 A bit backend heavy for me so take what I've said with a pinch of salt.

Seems to be working, only thing I did notice is that the named tags only seems to be pulling in some of them?

image
image

The `discount:black friday` one is added through that NA Bulk Price editor which Marketing will be using to set discount banners/badges etc. So I dunno if that has something to do with it?

**Review by hi-im-si — COMMENTED** [review]
My tests all seem to be coming back successfully.
Test details can be found [here](https://www.notion.so/openstudycollege/Sanity-Algolia-Webhook-Testing-b1191b281a2e49cd89812ab4935fa4e2?pvs=4)

Nice work man! Think we may be nearly there 🤞🏼

**Comment by mattwendzina** [comment]
> Blimey mate well done 😅 A bit backend heavy for me so take what I've said with a pinch of salt.
> 
> Seems to be working, only thing I did notice is that the named tags only seems to be pulling in some of them?
> 
> image image
> 
> The `discount:black friday` one is added through that NA Bulk Price editor which Marketing will be using to set discount banners/badges etc. So I dunno if that has something to do with it? That's on https://admin.shopify.com/store/openstudydev/products/7794836504831

As discussed, this is because namedTags is just an object so if there is more than one tag with the same key it will overwrite the old one. Hopefully this won't cause us any issues on the assumption of them just having one sale on a product..!

File: packages/osc-ecommerce/app/routes/actions/sanityAlgolia.tsx [hunk 1]
```diff
@@ -70,183 +106,549 @@ export const action: ActionFunction = async ({ request }: ActionArgs) => {
...
+            statusCode: 200,
+            body: 'ok',
+            result: 'No updates necessary',
+        };
+    }
+
+    // Get Shopify Storefront
+    const { storefront } = context;
+    // Get Shopify Admin API
+    const { adminApi } = context;
```
[the remark this answers is not in the document]
Comment (mattwendzina) [reply packages/osc-ecommerce/app/routes/actions/sanityAlgolia.tsx]: Done

File: packages/osc-ecommerce/app/routes/actions/sanityAlgolia.tsx [hunk 2]
```diff
@@ -70,183 +106,549 @@ export const action: ActionFunction = async ({ request }: ActionArgs) => {
...
-    const limitContent = (content: any) => {
-        return content.filter((_res: any, i: number) => {
-            if (i <= 10) {
-                return true;
+    // Get Sanity Client - Set to true to ensure we get the freshest data to keep algolia in sync
+    const sanityClient = getClient(true);
+
+    // Check sanity client is configured correctly
+    try {
+        await sanityClient.fetch('*[_type == "page"][0..9]');
```
[the remark this answers is not in the document]
Comment (mattwendzina) [reply packages/osc-ecommerce/app/routes/actions/sanityAlgolia.tsx]: Good shout, done.

File: packages/osc-ecommerce/app/routes/actions/sanityAlgolia.tsx [hunk 3]
```diff
@@ -48,12 +80,16 @@
...
     const jsonBody = JSON.parse(readableBody);
 
     const ids: Record<string, any> = {};
 
     // Ids from the request body have ids or are null if no change. webhookSync appears
-    // to expect an empty array if no changes rather than null - ie [] not [null]
+    // to expect an empty array if no changes rather than array with null - ie [] not [null]
     // This loop finds ones that are null and updates them to an empty array
     for (const key of Object.keys(jsonBody)) {
         if (key === 'ids') {
```
[the remark this answers is not in the document]
Comment (mattwendzina) [reply packages/osc-ecommerce/app/routes/actions/sanityAlgolia.tsx:95]: Happy with this hardcoded. Should be sufficient for our use cases.

**Review by stevenbister — APPROVED** [review]
Looks good from my point of view

**Pull request merged**